### PR TITLE
Rtc mix mode  with subseconds param. expressed in milliseconds

### PR DESCRIPTION
**Pull request by FRASTM**
stm32duino RTC add a parameter to set binary or bcd or Mix mode

The MIX mode of the RTC combines the binary and the BCD calendar modes
"The SSR binary down-counter is extended to 32-bit length and is free running. The time and
date calendar BCD registers are also available."

This is available of several stm32 devices like the stm32WL, stm32L4plus, stm32U5

The STM32RTC APi has subsecond parameter which is a value in **milliseconds**.
Because of the BINARY MIX mode, the parameter subsecond could be confusing as the subsecond register of the RTC is:

- a nb of millisec [0-999] in BCD mode

- a nb of ticks in BIN or MIX mode -> one tick is one counter unit at the fqce = RTC clock source freq / fqce_apre

For example, in MIX mode, when RTC is clocked by the LSE (32768Hz) the tick is downcounting at 256Hz
(1 tick is ~ 3.9ms )
The RTC subsecond register is converted in ms by  (0xFFFFFFFF - SubSecReg)  * 1000/256
And N miliseconds is converted in ticks (count unit) by 0xFFFFFFFF - (SubSecReg  * 256) /1000

With this PR, the STM32RTC API is always giving the subsecond parameter in a nb of milliseconds.
Especially with functions:

    getSubSeconds() and getTime()

    getAlarmSubSeconds()

    setAlarmSubSeconds() and setAlarmTime()

    setSubSeconds() and setTime() (but Subsecond value is ignored there RO register)

### CODE REVIEW

**Comment by FRASTM** [comment]
Testing with arduino RTC sketch when the on th nucleo_wl55jc or disco_l4s5i_iot1a
`  rtc.setRTCMode(STM32RTC::MODE_MIX);`
before `
  rtc.begin(); // initialize RTC 24H format`

File: src/rtc.c [hunk 1]
```diff
@@ -810,16 +810,8 @@ void RTC_StartAlarm(alarm_t name, uint8_t day, uint8_t hours, uint8_t minutes, u
...
-  if (name == ALARM_B) {
-    __HAL_RTC_ALARM_CLEAR_FLAG(&RtcHandle, RTC_FLAG_ALRBF);
-  } else
-#endif
-  {
-    __HAL_RTC_ALARM_CLEAR_FLAG(&RtcHandle, RTC_FLAG_ALRAF);
-  }
-  /* Disable the Alarm A interrupt */
+  /* Clear RTC Alarm Flag : done in the HAL_RTC_DeactivateAlarm function */
+  /* Disable the Alarm  interrupt */
```
Comment (fpistm) [inline src/rtc.c:814]: Seems not the case for all series. So it should be fine to keep this

**Review by fpistm — CHANGES_REQUESTED** [review]

File: src/rtc.c (hunk 1 above)
Comment (FRASTM) [reply src/rtc.c]: restored

**Comment by FRASTM** [comment]
The change to include the RTC MIX mode when manipulating the subsecond register
With this change, we can program alarm in MIX mode for few ms or seconds 
```
13:03:53.626 -> start at subsec = 0x12
13:03:53.626 -> set aAlarm A at = 0xda 
13:03:53.626 -> set aAlarm B at = 0x1012 --> 4096 is 16 seconds with a fqce_APRE
13:03:54.241 -> Alarm A Match! at subsec = 0xda
13:04:09.460 -> Alarm B Match! at subsec = 0x1012
```

File: src/rtc.c [hunk 2]
```diff
@@ -758,6 +807,36 @@ void RTC_StartAlarm(alarm_t name, uint8_t day, uint8_t hours, uint8_t minutes, u
     UNUSED(mask);
 #endif /* !STM32F1xx */
 
+    /* Set RTC_Alarm */
+    HAL_RTC_SetAlarm_IT(&RtcHandle, &RTC_AlarmStructure, RTC_FORMAT_BIN);
+    HAL_NVIC_SetPriority(RTC_Alarm_IRQn, RTC_IRQ_PRIO, RTC_IRQ_SUBPRIO);
+    HAL_NVIC_EnableIRQ(RTC_Alarm_IRQn);
+  } else {
+    /* We have an SubSecond alarm to set in RTC_BINARY_MIX mode */
```
Comment (fpistm) [inline src/rtc.c:815]: Seems nothing indicate we are in binary mode, only that arguments is not correct to start calendar alarm.

File: src/rtc.c [hunk 3]
```diff
@@ -758,6 +807,36 @@ void RTC_StartAlarm(alarm_t name, uint8_t day, uint8_t hours, uint8_t minutes, u
...
+      if (name == ALARM_B) {
+        /* Expecting RTC_ALARMSUBSECONDBINMASK_NONE for the subsecond mask on ALARM B */
+        RTC_AlarmStructure.AlarmSubSecondMask = mask << RTC_ALRMBSSR_MASKSS_Pos;
+      } else
+#endif
+      {
+        /* Expecting RTC_ALARMSUBSECONDBINMASK_NONE for the subsecond mask on ALARM A */
+        RTC_AlarmStructure.AlarmSubSecondMask = mask << RTC_ALRMASSR_MASKSS_Pos;
+      }
+      /* Special case for ALARM B configuration for stm32WL Lorawan */
```
Comment (fpistm) [inline src/rtc.c:833]: why only for STM32W?

File: src/rtc.c (hunk 2 above)
Comment (FRASTM) [reply src/rtc.c:815]: + checking the RTC mode is binary with 
`else if (RtcHandle.Init.BinMode != MODE_BINARY_NONE) {`

**Comment by FRASTM** [comment]
Add the sketch **mixRTCAlarm**

**Comment by FRASTM** [comment]
using the https://github.com/stm32duino/STM32RTC/pull/98
with MSP Init function  for any stm32 device with a SS underflow flag
with BCD mode for getTime and GetAlarm

in BIN (BINARY ONLY) mode,  the SSR is down-counting ck_apre ticks (like the MIX mode does) on 32-bit and Timer/Date Registers are not used
in BCD (BINARY NONE) mode,  the SSR is down-counting ck_apre ticks on 16-bit

**Comment by FRASTM** [comment]
Testing on stm32g474 nucleo : RTC has BCD mode only or stm32wl55 nucleo  RTC in BCD or MIX mode:
```
12:00:27.195 -> Start at 16:22:06.300
12:00:27.195 -> Set Alarm A in 12s (at 16:22:18)
12:00:27.195 -> Set Alarm B (in 600 ms) at 902 ms
12:00:27.806 -> Alarm B Match at 902 ms
12:00:39.893 -> Alarm A Match at 16:22:18
```
Testing

**Review by fpistm — APPROVED** [review]
LGTM
Tested with https://github.com/stm32duino/STM32LoRaWAN/pull/25

**Pull request merged**